_rules/SC4-1-2-button-has-name.md

```diff
@@ -0,0 +1,137 @@
+---
+name: Buttons have an accessible name
+
+description: |
+  Each button element has an accessible name
+
+success_criterion:
+- 4.1.2
+
+test_aspects:
+- DOM Tree
+- CSS Styling
+
+authors:
+- Wilco Fiers
+- Stein Erik Skotkjerra
+---
+
+## Test procedure
+
+### Applicability
+
+The rule applies to elements that are [visible on the page](#visible-on-the-page) or [exposed to assistive technologies](#exposed-to-assistive-technologies) with the [semantic role](#semantic-role) of `button`, except for `input` elements of `type="image"`.
+
+### Expectation
+
+Each target element has an [accessible name](#accessible-name) that is [non-empty](#non-empty)
+
+## Assumptions
+
+- The rule assumes that all buttons are user interface components as defined by WCAG 2.
+
+## Accessibility Support
+
+There are no major accessibility support issues known for this rule.
+
+## Background
+
+- [https://www.w3.org/TR/UNDERSTANDING-WCAG20/ensure-compat-rsv.html](https://www.w3.org/TR/UNDERSTANDING-WCAG20/ensure-compat-rsv.html)
+- [https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=412#qr-ensure-compat-rsv](https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=412#qr-ensure-compat-rsv)
+- [ARIA14: Using aria-label to provide an invisible label where a visible label cannot be used](https://www.w3.org/TR/WCAG20-TECHS/ARIA14.html)
+- [https://www.w3.org/TR/WCAG20-TECHS/ARIA16.html](https://www.w3.org/TR/WCAG20-TECHS/ARIA16.html)
+
+# Test cases
+
+## Passed
+
+```html
+<!-- Regular button -->
+<button>My button</button>
+```
+
+```html
+<!-- Value attribute as the accessible name -->
+<input type="submit" value="Submit">
+```
+
+```html
+<!-- aria-label for the accessible name -->
+<button aria-label="My button"></button>
+```
+
+```html
+<!-- span tag with role button and has name defined by aria-label -->
+<span role="button" aria-label="My button"></button>
+```
+
+```html
+<!-- summary element has a default sematic role of button -->
+<summary>Press Here</summary>
+```
+
+```html
+<!-- disabled elements are also applicable -->
+<button disabled>Delete</button>
+```
+
+```html
+<html>
+  <style>
+    .notInPage {
+      position: absolute;
+      left: -9999px;
+      top: -9999px;
+    }
+  </style>
+  <body>
+    <!-- Off screen elements should be tested -->
+    <button class='notInPage'>Save</button>
+  </body>
+</html>
+```
+
+## Failed
+
```

Comment:
Add one failed test case of an off-screen element without an accessible name

Reply:
Sorted.